[mmaimer33] iP

mmaimer33 opened the pull request:
## PixlBot
This is a command-line bot to help maintain tasks, built on the Duke codebase found at the [CS2103 iP repo](https://github.com/nus-cs2103-AY2223S2/ip).

#### List of commands
- list
- todo `<description>`
- deadline `<description>` /by `<date>`
- event `<description>` /from `<date/time>` /to `<date/time>`
- mark `<task-number>`
- unmark `<task-number>`
- delete `<task-number>`
- find `<keyword>`
- bye

#### Typical flow of chatbot
1. Run the program (current version: *.jar* file -**v.0.1-alpha**)
2. Program will find and load existing list or create a new one
3. Use commands to add/edit/remove/find tasks
4. Enter `bye` to exit
5. Program will save the current list
6. Program ends.

The main program is called `Duke.java`, with a main method looking like this:
```java
public static void main(String[] args) {
    new Duke("data/Duke").run();
}
```
The remainder can more-or-less be traced from here.

#### How is it?
Good! It can help you maintain tasks:
- [x] ToDos (just a plain task)
- [ ] Deadlines
- [ ] Events that
- [x] May be complete

A review:
> Very good program. Extremely useful and helped me pass every single exam I have ever taken in my life 👍🏻. *- Creator of the program <sub>(may be biased)</sub>*

Commit:
Add Gradle support

mmaimer33 committed:
Implemented basic echo functionality.

mmaimer33 committed:
Changed chat format to use colors. Refactored PixlBot printing.

mmaimer33 committed:
Moved final values to new Values class.

mmaimer33 committed:
Implemented list functionality.

mmaimer33 committed:
Refactored code to use a new Task class.

mmaimer33 committed:
Refactored code to use separate functions for each command.

mmaimer33 committed:
Added command: mark (mark task as done).

mmaimer33 committed:
Added command: unmark (mark task as not done).

mmaimer33 committed:
Added Task subclasses ToDo, Deadline, Event.

mmaimer33 committed:
Added commands: todo, deadline, event. Implemented new task system.

mmaimer33 committed:
Fixed task add commands to accept multiple words.

mmaimer33 committed:
Created automated test.

mmaimer33 committed:
Implemented error handling using a new DukeException.

mmaimer33 committed:
Added command: delete.

mmaimer33 committed:
Added a message showing number of tasks left for add and delete operations.

mmaimer33 committed:
Update bot to use switch/case for efficiency.

mmaimer33 committed:
Implement saving and loading list data.

Data is saved to 'src/data/duke' and loaded from the same location if it
exists.

mmaimer33 committed:
Change names to better follow Java standards (minor).

mmaimer33 committed:
Convert Deadline dueDate field to java.time.LocalDate

mmaimer33 committed:
Merge branch 'branch-level-8'

# Conflicts:
#	src/main/java/Duke.java

mmaimer33 committed:
Create new class Ui to handle user interactions

mmaimer33 committed:
Create new class TaskList to handle list operations.

mmaimer33 committed:
Create new class Storage to handle hard disk storage and retrieval.

mmaimer33 committed:
Create Parser class to handle command re-direction +
Create Command classes for all existing commands

Create abstract class Command with subclasses for each of the existing
commands. Implementation not complete.

mmaimer33 committed:
Refactor Duke to use new utility classes

mmaimer33 committed:
Implement Command subclasses + Fix Storage

* The Command subclasses all have an empty execute method and do nothing.
* The Storage class stores data as a TaskList, which has issues with
serialisation.

* Implement the execute method of each Command in accordance to original
behaviour.
* Change Storage class to store data as ArrayList<Task>.

mmaimer33 committed:
Divide classes into packages

All files are in the same src/java directory. Move them into one of:
* common package duke
* duke.task
* duke.command

mmaimer33 committed:
Merge branch 'add-gradle-support' of github.com:mmaimer33/ip

mmaimer33 committed:
Add JUnit tests

Adding tests required adding equals() methods to some classes.

mmaimer33 committed:
(Minor) Replace an assertEquals with assertNull

mmaimer33 committed:
Add jar file support

Changed default storage location to data folder at same level as src

mmaimer33 committed:
Add JavaDoc comments and headers.

mmaimer33 committed:
Refactor some code to follow Java coding standard

chuakid commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,75 @@
+    }
+
+    /**
+     * Utility function to find index of element in a String array.
+     * @param arr The array to search.
+     * @param item The item to find.
+     * @return Index of item if present; -1 otherwise.
+     */
```
I like the effort in giving every class and function a javadoc. However, should the description be in this form?

chuakid commented on src/main/java/duke/command/DeadlineCommand.java:
```diff
@@ -0,0 +1,54 @@
+        }
+        // Get due date.
+        StringBuilder dueDate = new StringBuilder();
+        for (int i = byIndex+1; i < parts.length; i++) {
+            dueDate.append(i == byIndex+1 ? "" : Values.SPACE);
```
Operators here could benefit from some whitespace.

chuakid commented on src/main/java/duke/command/DeleteCommand.java:
```diff
@@ -0,0 +1,23 @@
+    public void execute(Ui ui, TaskList list, String command) throws DukeException {
+        try {
+            Task task = list.removeTask(Integer.parseInt(command.split("\\s+")[1]) - 1);
+            ui.pixlPrint("Removed the task:\n" +
+                    "\t" + task.formatTask() +
+                    "\nYou now have " + list.getSize() + " task(s) in the list.");
```
I like the line-breaks used here.

chuakid reviewed (commented):
Mostly follows coding standards with some minor nits. Good job!

mmaimer33 committed:
Add command: find

Also refactored command parsing to use a constant for regex.

mmaimer33 committed:
Merge branch 'branch-A-JavaDoc'

mmaimer33 committed:
Merge branch 'branch-A-CodingStandard'

mmaimer33 committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/duke/command/DeadlineCommand.java
#	src/main/java/duke/command/EventCommand.java

mmaimer33 committed:
Implement checkstyle and edit files to follow Java style conventions

mmaimer33 committed:
Modify files further to match checkstyle conventions

mmaimer33 committed:
Merge branch 'branch-A-CheckStyle'

kenzantonius reviewed (commented):
Great job! I do not find anything wrong in the Code Quality. Keep up the good work! 

One good thing to note: You really did a good job on splitting a code in a line into some line.

mmaimer33 committed:
Setup javafx

mmaimer33 committed:
Create basic un-interactive GUI

mmaimer33 committed:
Refactor code to use GUi for interaction

mmaimer33 committed:
Merge branch 'branch-Level-10'

mmaimer33 committed:
Delete extra file Launcher.java

mmaimer33 committed:
Use assert to ensure GUi is functional

mmaimer33 committed:
Merge pull request #2 from mmaimer33/branch-A-Assertions

Implemented assertions.

mmaimer33 committed:
Improve code quality to match course conventions

Some methods are too long (especially execute methods in the add task
commands). Some method names are too vague. Code readability is not
great occasionally.

Let's extract out the task->execute functionality to parse the command
and get relevant details into separate methods. Break up and rename some
methods. Improve spacing of method bodies to be consistent.

This makes the code more readable and makes debugging easier.

mmaimer33 committed:
Fix bug: application cannot start.

The fields of GUi are not initialized and cause a runtime error that
prevents the app from opening.

Let's fix it by calling the initializeGUiFields() method.

Init code was moved to the method and calling it was necessary.

mmaimer33 committed:
Merge pull request #3 from mmaimer33/branch-A-CodeQuality

Merge branch-A-CodeQuality as Pull Request.

mmaimer33 committed:
Add support for tags in Task and Command

Tagging not yet fully implemented.

mmaimer33 committed:
Temp commit

mmaimer33 committed:
Add tags implementation to Deadline and Event commands

mmaimer33 committed:
Merge branch 'branch-C-Tagging'

mmaimer33 committed:
Fix style errors according to checkstyle.xml

mmaimer33 committed:
Make small Ui changes, fix logo

mmaimer33 committed:
Update README to contain actual user guide
